x/blob/types/payforblob.go

```diff
@@ -331,6 +332,51 @@ func ValidateBlobs(blobs ...*Blob) error {
 	return nil
 }
 
+// validateBlobData returns an error if the size of data is zero or too large.
+func validateBlobData(appVersion uint64, data []byte) error {
+	if len(data) == 0 {
+		return ErrZeroBlobSize
+	}
+
+	maxBlobSize := blobSizeUpperBound(appVersion)
+	if len(data) > maxBlobSize {
+		return ErrBlobSizeTooLarge.Wrapf("max blob size is %d bytes", maxBlobSize)
+	}
+	return nil
+}
+
+// blobSizeUpperBound returns an upper bound for the max valid blob size based
+// on the upper bounds for square size and block bytes. Note it is possible that
+// blobs of this size do not fit in a block because the limiting factor may be a
+// parameter in application state (i.e. GovMaxSquareSize). Additionally, even if
+// the application state parameters are at their upper bounds, the number of
+// shares available for blob bytes may be less than estimated in this function
+// (i.e. if the PFB tx shares occupy more than one share). As a result, the
+// upper bound returned by this function may over-estimate the max valid blob
+// size but it should not under-estimate. Consequently, it may be used to
+// immediately reject blobs that are too large but blobs smaller than this upper
+// bound may still fail to be included in a block.
+func blobSizeUpperBound(appVersion uint64) int {
+	maxSquareSize := appconsts.SquareSizeUpperBound(appVersion)
+	maxShares := maxSquareSize * maxSquareSize
+
+	// Subtract one from maxShares because at least one share must be occupied
+	// by the PFB tx associated with this blob.
+	maxBlobShares := maxShares - 1
+	maxBlobBytes := maxBlobShares * appconsts.ContinuationSparseShareContentSize
+
+	return min(maxBlobBytes, coretypes.MaxBlockSizeBytes)
+}
```

Comment:
I think we should simplify this upper bound to compute the max number of bytes in a data square based on app version and reject based on that.